src/base/torrentfileswatcher.cpp

```diff
@@ -0,0 +1,618 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentfileswatcher.h"
+
+#include <QtGlobal>
+#include <QDir>
+#include <QDirIterator>
+#include <QFile>
+#include <QFileSystemWatcher>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+#include <QSaveFile>
+#include <QTextStream>
+#include <QThread>
+#include <QTimer>
+#include <QVariant>
+
+#include <chrono>
+
+#include "base/algorithm.h"
+#include "base/bittorrent/magneturi.h"
+#include "base/bittorrent/torrentcontentlayout.h"
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/torrent.h"
+#include "base/bittorrent/torrentinfo.h"
+#include "base/exceptions.h"
+#include "base/global.h"
+#include "base/logger.h"
+#include "base/profile.h"
+#include "base/settingsstorage.h"
+#include "base/tagset.h"
+#include "base/utils/fs.h"
+#include "base/utils/string.h"
+
+using namespace std::chrono_literals;
+
+const std::chrono::duration WATCH_INTERVAL = 10s;
+const int MAX_FAILED_RETRIES = 5;
+const QString CONF_FILE_NAME {QStringLiteral("watched_folders.json")};
+
+const QString OPTION_ADDTORRENTPARAMS {QStringLiteral("add_torrent_params")};
+const QString OPTION_RECURSIVE {QStringLiteral("recursive")};
+
+const QString PARAM_CATEGORY {QStringLiteral("category")};
+const QString PARAM_TAGS {QStringLiteral("tags")};
+const QString PARAM_SAVEPATH {QStringLiteral("save_path")};
+const QString PARAM_OPERATINGMODE {QStringLiteral("operating_mode")};
+const QString PARAM_STOPPED {QStringLiteral("stopped")};
+const QString PARAM_CONTENTLAYOUT {QStringLiteral("content_layout")};
+const QString PARAM_AUTOTMM {QStringLiteral("use_auto_tmm")};
+const QString PARAM_UPLOADLIMIT {QStringLiteral("upload_limit")};
+const QString PARAM_DOWNLOADLIMIT {QStringLiteral("download_limit")};
+const QString PARAM_SEEDINGTIMELIMIT {QStringLiteral("seeding_time_limit")};
+const QString PARAM_RATIOLIMIT {QStringLiteral("ratio_limit")};
+
+namespace
+{
+    TagSet parseTagSet(const QJsonArray &jsonArr)
+    {
+        TagSet tags;
+        for (const QJsonValue &jsonVal : jsonArr)
+            tags.insert(jsonVal.toString());
+
+        return tags;
+    }
+
+    QJsonArray serializeTagSet(const TagSet &tags)
+    {
+        QJsonArray arr;
+        std::transform(tags.cbegin(), tags.cend(), std::back_inserter(arr)
```

Comment:
Don't you always suggest using algorithms instead of loops?
What would be the advantage of a loop here? You could have mentioned it right away, instead of waiting for the next round of discussion. I certainly don't mind talking to you, but it can be a bit unproductive when the delay between iterations of our discussion reaches a day (or more).

Reply:
If you're just wondering, I don't have a clear preference for what to use in such trivial cases, so I chose the standard algorithm, bearing in mind that you often asked someones to use one of them instead of a loop.

Reply:
Use of the algorithm here better expresses the intent IMO. Unless using a for-loop here brings some big advantage (e.g performance or whatever), this way is better.

Reply:
>so I chose the standard algorithm, bearing in mind that you often asked someones to use one of them instead of a loop.

Thanks for the consideration.

I admit the comment on this code is personal opinion.
```c++
for (const QString &tag : tags)
    arr.append(QJsonValue(tag));  // maybe the `QJsonValue` constructor can be omitted?
```
Compared to the simple for-loop above the std::transform is harder to read and unnecessary long and we don't need some special "execution policy" here, so most of the time I would go with for-loop instead.
The `std::back_inserter` is another thing that I wouldn't expect to see in non-generic algorithm code. It is good at abstracting the container type away but here we already have a container that we are going to use so why the abstraction? Using it only create harder to read code.